### ISSUE TITLE
internal error: entered unreachable code when using tokio integration

### ISSUE THREAD

**Issue by netthier** [bk-rs/ssh-rs#28]
I'm using an `AsyncSession<tokio::net::TcpStream>` concurrently across multiple tokio tasks to open channels and execute commands over them.
However, I noticed some of those tasks crashing due to this error:
```rust
thread 'tokio-runtime-worker' panicked at 'internal error: entered unreachable code: ', /home/nett/.cargo/registry/src/github.com-1ecc6299db9ec823/async-ssh2-lite-0.4.1/src/session_stream/impl_tokio.rs:39:21
```
or this one
```rust
thread 'tokio-runtime-worker' panicked at 'internal error: entered unreachable code: ', /home/nett/.cargo/registry/src/github.com-1ecc6299db9ec823/async-ssh2-lite-0.4.1/src/session_stream/impl_tokio.rs:81:17
```
What is going on here?

**Comment by netthier** [comment]
The backtraces always point to any async_ssh2_lite function being invoked, e.g.
```
thread 'tokio-runtime-worker' panicked at 'internal error: entered unreachable code: ', /home/nett/.cargo/registry/src/github.com-1ecc6299db9ec823/async-ssh2-lite-0.4.1/src/session_stream/impl_tokio.rs:39:21
stack backtrace:
   0: rust_begin_unwind
             at /rustc/659e169d37990b9c730a59a96081f2ef7afbe8f1/library/std/src/panicking.rs:575:5
   1: core::panicking::panic_fmt
             at /rustc/659e169d37990b9c730a59a96081f2ef7afbe8f1/library/core/src/panicking.rs:64:14
   2: async_ssh2_lite::session_stream::impl_tokio::<impl async_ssh2_lite::session_stream::AsyncSessionStream for tokio::net::tcp::stream::TcpStream>::x_with::{{closure}}
             at /home/nett/.cargo/registry/src/github.com-1ecc6299db9ec823/async-ssh2-lite-0.4.1/src/session_stream/impl_tokio.rs:39:21
   3: <core::pin::Pin<P> as core::future::future::Future>::poll
             at /rustc/659e169d37990b9c730a59a96081f2ef7afbe8f1/library/core/src/future/future.rs:125:9
   4: async_ssh2_lite::session_stream::AsyncSessionStream::rw_with::{{closure}}
             at /home/nett/.cargo/registry/src/github.com-1ecc6299db9ec823/async-ssh2-lite-0.4.1/src/session_stream/mod.rs:41:9
   5: <core::pin::Pin<P> as core::future::future::Future>::poll
             at /rustc/659e169d37990b9c730a59a96081f2ef7afbe8f1/library/core/src/future/future.rs:125:9
   6: async_ssh2_lite::session::AsyncSession<S>::sftp::{{closure}}
             at /home/nett/.cargo/registry/src/github.com-1ecc6299db9ec823/async-ssh2-lite-0.4.1/src/session.rs:392:13
   7: ssh_worker::ssh::Client::new_sftp::{{closure}}
             at /home/nett/Stuff/sisyphos/ssh-worker/src/ssh.rs:44:31
```
I'm not sure, but it seems to happen when I call multiple methods like that concurrently on the same `AsyncSession`

**Comment by vkill** [comment]
Hi, @netthier
Please provide an example. Thank you.

**Comment by netthier** [comment]
I came up with a reproducing example, but it seems to work less reliably than my actual codebase.
Running it a few times in release mode should produce the errors though:
```
thread 'tokio-runtime-worker' panicked at 'internal error: entered unreachable code: ', /home/nett/.cargo/registry/src/github.com-1ecc6299db9ec823/async-ssh2-lite-0.4.1/src/session_stream/impl_tokio.rs:39:21
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

https://gist.github.com/42aadced5dea7d9eab3ba1d3d9b8c4d0

Given the almost "random" occurrences of this error and it only occurring when using a Session concurrently, it seems like some kind of race condition?

Also, the code sometimes throws this error as well:
```
ssh-bug: libssh2/src/transport.c:349: _libssh2_transport_read: Assertion `remainbuf >= 0' failed.
thread 'tokio-runtime-worker' panicked at 'called `Result::unwrap()` on an `Err` value: Ssh2(Error { code: Session(-1), msg: "Failed waiting for channel success" })

Location:
    src/main.rs:54:9', [1]    15413 IOT instruction (core dumped)
```

**Comment by vkill** [comment]
Hi @netthier, I add [tests](https://github.com/bk-rs/ssh-rs/blob/main/async-ssh2-lite/tests/integration_tests/tokio_spawn_session.rs#L44) and run `SSH_SERVER_HOST=1.1.1.1 SSH_SERVER_PORT=22 SSH_USERNAME=root SSH_PRIVATEKEY_PATH=~/.ssh/id_rsa cargo test -p async-ssh2-lite --features _integration_tests,_integration_tests_tokio_ext,async-io,tokio -- integration_tests::tokio_spawn_session --nocapture` , not crash happened.

**Comment by vkill** [comment]
But I tend to use [bb8-async-ssh2-lite](https://github.com/bk-rs/ssh-rs/blob/main/bb8-async-ssh2-lite/demo/src/tokio_tcp_stream.rs) in concurrent case.

**Comment by netthier** [comment]
Thanks for the response, this bug does seem to be difficult to trigger.
Using bb8 fixed the bug in my main project, but I was still able to trigger it exactly once in a test project (but not anymore after that).
Seems like its related to lots of channels being opened at the same time in the same session?

**Comment by vkill** [comment]
Yes, theoretically,  should exist one channel in one session simultaneously.

**Comment by netthier** [comment]
> Yes, theoretically, should exist one channel in one session simultaneously.

@vkill Is this documented somewhere? Both this and the fact that you can't call `exec` on a channel twice are pitfalls I fell into :thinking: 
If not, it should be

**Comment by netthier** [comment]
I'm still getting the panic with your bb8 crate under very high load. I'll try to write a better reproducer over the course of the week, as my program does a ton of stuff over SSH.

**Comment by vkill** [comment]
@netthier Please try change [here](https://github.com/bk-rs/ssh-rs/blob/main/async-ssh2-lite/src/session_stream/impl_tokio.rs#L38-L40) to `BlockDirections::None => continue,` , because I found this [commit](https://github.com/andras-pinter/ssh-rs/commit/80fd5bb116c2e6850af666983ce19f793e39a4d8)

**Comment by netthier** [comment]
Thank you for finding this, seems like it has fixed the issue. I have not been able to trigger this bug in my project anymore.
I created a PR containing the patch here: https://github.com/bk-rs/ssh-rs/pull/31

**Comment by netthier** [comment]
Seems like it's not that easy after all. Instead of panicking, acquiring a session now errors out sometimes:
```
Err(
        Ssh2(
            Error {
                code: Session(
                    -21,
                ),
                msg: "Unable to startup channel",
            },
        ),
    )
```
However, handling this kind of error is still way easier than catching a panic.

It seems like this error is related to too many concurrent channels on a single session, meaning our suspicion was correct: https://sysprogs.com/w/forums/topic/ssh2-warnings-server-becomes-locked-up/
However, this still occurs with your bb8 crate. I haven't used bb8 before, does your crate limit the number of channels per session?

**Comment by netthier** [comment]
Do Channels automatically close themselves on `Drop`? I wasn't able to find any custom `Drop` impls for it.
Is it possible that a Channel stays open for a short while after being dropped, with bb8 thinking the session can be reused for new channels?

EDIT: I switched out all occurrences of `pool.get()` to `pool.dedicated_connection()` to check this. While I didn't get any `-21` errors anymore, I'm now getting
```
    Err(
        HandshakeError(
            Ssh2(
                Error {
                    code: Session(
                        -13,
                    ),
                    msg: "Failed getting banner",
                },
            ),
        ),
    ),
```
which I'd assume is related to a different limit being reached (MaxStartups in sshd_config maybe?)

I believe a solution here would be to use something like a semaphore to limit the maximum amount of concurrent, unauthenticated connections in a bb8 connection pool, and another semaphore to limit the maximum amount of channels in a session.
Both of these need to be limited somehow, as ssh servers seem to have hard limits on those.

**Comment by netthier** [comment]
Increasing `MaxStartups` in my `sshd_config` fully fixed this issue, meaning a connection limit support is the way forward here.
I'll open a new issue regarding that, and this issue can be closed once the PR is merged.

**Comment by vkill** [comment]
> Do Channels automatically close themselves on Drop? 
Yes, https://docs.rs/ssh2/0.9.3/src/ssh2/channel.rs.html#488 and https://www.libssh2.org/libssh2_channel_free.html
But I suggest `channel.close().await`

**Comment by vkill** [comment]
In short, I like one channel in one session simultaneously, close the channel actively, and then use bb8 `max_size` to limit the number of sessions.

**Comment by netthier** [comment]
> > Do Channels automatically close themselves on Drop? 
>  Yes, https://docs.rs/ssh2/0.9.3/src/ssh2/channel.rs.html#488 and https://www.libssh2.org/libssh2_channel_free.html
>  But I suggest `channel.close().await`

Ah, I overlooked that. I'll need to check closer if it's possible that multiple channels exist concurrently per session.

**Comment by vkill** [comment]
v0.4.2

**Comment by vkill** [comment]
https://github.com/bk-rs/ssh-rs/blob/b79d97f5653695dc4b9242635476d400023f4d3d/async-ssh2-lite/tests/integration_tests/tokio_spawn_session.rs#L57-L61

Maybe `Ssh2(Error { code: Session(-4), msg: "Unexpected error" }) ` when open many channels in one session simultaneously.
